cmd/tempo/main.go

```diff
@@ -171,3 +185,110 @@ func loadConfig() (*app.Config, error) {
 
 	return config, nil
 }
+
+func installOpenTracingTracer(config *app.Config) (func(), error) {
+	level.Info(log.Logger).Log("msg", "initialising OpenTracing tracer")
+
+	// Setting the environment variable JAEGER_AGENT_HOST enables tracing
+	trace, err := tracing.NewFromEnv(fmt.Sprintf("%s-%s", appName, config.Target))
+	if err != nil {
+		return nil, errors.Wrap(err, "error initialising tracer")
+	}
+	return func() {
+		if err := trace.Close(); err != nil {
+			level.Error(log.Logger).Log("msg", "error closing tracing", "err", err)
+			os.Exit(1)
+		}
+	}, nil
+}
+
+func installOpenTelemetryTracer(config *app.Config) (func(), error) {
+	level.Info(log.Logger).Log("msg", "initialising OpenTelemetry tracer")
+
+	// for now, migrate OpenTracing Jaeger environment variables
+	migrateJaegerEnvironmentVariables()
+
+	exp, err := jaeger.New(jaeger.WithCollectorEndpoint())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create Jaeger exporter")
+	}
+
+	resources, err := resource.New(context.Background(),
+		resource.WithAttributes(
+			semconv.ServiceNameKey.String(fmt.Sprintf("%s-%s", appName, config.Target)),
+			semconv.ServiceVersionKey.String(build.Version),
+		),
+		resource.WithHost(),
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to initialise trace resuorces")
+	}
+
+	tp := tracesdk.NewTracerProvider(
+		tracesdk.WithBatcher(exp),
+		tracesdk.WithResource(resources),
+	)
+	otel.SetTracerProvider(tp)
+
+	shutdown := func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```

Comment:
What is the rationale for 5 second timeout on shutdown?

Reply:
This used in the Jaeger example as well: https://github.com/open-telemetry/opentelemetry-go/blob/23cb9396117ee0be80981ec9a60d1d1620688e29/example/jaeger/main.go#L78-L83
I'm not sure if this is really needed, other examples just use `context.Background()`.